spring-kafka/src/main/java/org/springframework/kafka/retrytopic/SameIntervalTopicReuseStrategy.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2021-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.retrytopic;
+
+/**
+ *
+ * Strategy for topic reuse when multiple, sequential retries have the same backoff interval.
+ *
+ * @author João Lima
+ * @since 3.1
+ *
+ */
+public enum SameIntervalTopicReuseStrategy {
```

Comment:
Initially I had modeled something like `ExponentialBackoffMaxDelayTopicReuseStrategy`, but it would be too specific for exponential backoff. So I went with a more generic concept, with the logic in `DestinationTopicPropertiesFactory` also being more generic.

It turns out that, with this `SameIntervalTopicReuseStrategy` idea, the `FixedDelayStrategy` could be removed, because the logic implemented for `SameIntervalTopicReuseStrategy` itself would be enough to conclude that a single retry topic should be used also for fixed backoffs. Unfortunately, the feature went GA.

Maybe we may mark `FixedDelayStrategy` as deprecated, with it being removed and replaced by `SameIntervalTopicReuseStrategy` in some future major release.

Reply:
I think this makes total sense - fixed delays are only a specific case of the broader `topic resuse` concept. So IMHO having both concepts at the same time tend to be a bit confusing. I'm thinking we could deprecate the `FixedDelayStrategy` at once in this PR and use the new option. WDYT?

But for that to work, I think we'd need to adapt the `FixedDelayStrategy` part to respond to both configurations.

The part that called my attention is this one:
https://github.com/spring-projects/spring-kafka/blob/3f645b11a1a2f3afa6a8629d18cf83506174f0d1/spring-kafka/src/main/java/org/springframework/kafka/retrytopic/DestinationTopicPropertiesFactory.java#L114-L120

Maybe it should also look for the new configuration, and if it's `SINGLE_TOPIC` respect that regardless of the `FIXED` strategy?

This way we can deprecate the old enum and methods and users can rely only on the new one.

Reply:
> I'm thinking we could deprecate the FixedDelayStrategy at once in this PR and use the new option. WDYT?

@tomazfernandes, you are the bosses here. Feel free to direct me.

Reply:
Now that I read again, I think I mostly said what you'd already said 😄

Only adding that in order to deprecate the `FixedDelayStrategy`, which I believe we should do, we need to make the `SameIntervalTopicReuseStrategy` also work for fixed delays.

Reply:
I will wait for feedbacks from @garyrussell before actually changing things.

Reply:
I plan to look at this tomorrow; but since @tomazfernandes is the original author of this feature, I trust his words.

Reply:
I agree with deprecating the FDS, with the aim to remove it in 3.1.

Reply:
We should probably also add a disclaimer in the javadocs / documentation that `SameIntervalTopicReuseStrategy` doesn't work (at least yet) when the same interval is in the middle of the backoff values. Though the OOTB BackOff policies doesn't allow that, we've seen users create custom policies that would allow such cases.